### PR TITLE
Read package.json workspaces field and "yarn workspaces info" output

### PR DESCRIPTION
**Pull request by techtangents**
Rewrote the JSON decoding to use io-ts. This replaces a bunch of the functions in JsonUtils. io-ts seems much easier to use than what I had.

Simplified the PackageJson structure to make it easier to parse and serialize with io-ts.

Added ability to parse the "workspaces" field from package.json. Added functionality to run "yarn workspaces info" and parse the result. These will be used to build monorepo support.

### CODE REVIEW

File: src/test/ts/core/PackageJsonTest.ts [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+      });
+      assert.deepEqual(actual, E.right({
+        name: 'blah',
+        version: { major: 1, minor: 2, patch: 3, preRelease: 'rc', buildMetaData: undefined },
+        workspaces: [ 'cat', 'dog*' ],
+        fridge: 'donkey'
+      }));
+    });
+  });
+  describe('round-trip', () => {
```
Comment (lnewson) [inline src/test/ts/core/PackageJsonTest.ts:57]: Newline here maybe? Also, should these be `context` instead of describe?

**Review by lnewson — APPROVED** [review]

File: src/test/ts/core/PackageJsonTest.ts (hunk 1 above)
Comment (techtangents) [reply src/test/ts/core/PackageJsonTest.ts:58]: Newline added.

`context` and `describe` are the same thing. I don't think it matters. I think it reads ok: "describe the round-trip behaviour".

Comment (techtangents) [reply src/test/ts/core/PackageJsonTest.ts:58]: Having a look at https://mochajs.org/#bdd - it reads a bit differently to the grouping I have here.

File: src/test/ts/core/YarnWorkspacesTest.ts [hunk 2]
```diff
@@ -0,0 +1,26 @@
+import { describe, it } from 'mocha';
+import * as chai from 'chai';
+import * as chaiAsPromised from 'chai-as-promised';
+import * as YarnWorkspaces from '../../../main/ts/core/YarnWorkspaces';
+
+const assert = chai.use(chaiAsPromised).assert;
+
+describe('info', () => {
+  it('reads a test scenario with 2 modules', () => {
```
Comment (ashfordneil) [inline src/test/ts/core/YarnWorkspacesTest.ts:9]: Maybe add a "foo depends on bar" to this test so we can make sure that works?

**Review by ashfordneil — APPROVED** [review]

File: src/test/ts/core/YarnWorkspacesTest.ts (hunk 2 above)
Comment (techtangents) [reply src/test/ts/core/YarnWorkspacesTest.ts:9]: On the one hand: I don't want to have to test yarn. On the other hand: I kinda need to be sure it does what I want and a test breaks if it changes. I'll add it.

Comment (techtangents) [reply src/test/ts/core/YarnWorkspacesTest.ts]: Done.

**Pull request merged**